opentracing-api/src/main/java/io/opentracing/ScopeManager.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2016-2017 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.opentracing;
+
+/**
+ * The {@link ScopeManager} interface abstracts both the activation of {@link Span} instances (via
+ * {@link ScopeManager#activate(Span)}) and access to an active {@link Span}/{@link Scope}
+ * (via {@link ScopeManager#active()}).
+ *
+ * @see Scope
+ * @see Tracer#setScopeManager(ScopeManager)
+ */
+public interface ScopeManager {
+    /**
+     * Make a {@link Span} instance active.
+     *
+     * @param span the {@link Span} that should become the {@link #active()}
+     * @return a {@link Scope} instance to control the end of the active period for the {@link Span}. It is a
+     * programming error to neglect to call {@link Scope#close()} on the returned instance.
+     */
+    Scope activate(Span span);
+    Scope activate(Span span, Scope.Observer scopeObserver);
+
+    /**
+     * @return the currently active {@link Scope} which can be used to access the currently active
+     * {@link Scope#span()}
+     */
+    Scope active();
+
+    /**
+     * @return whether {@link Scope} instances support deferral by implementing the advanced
+     * {@link ContinuableScope} interface.
+     */
+    boolean canDefer();
```

Comment:
what would be an example of a scope that cannot be deferred? I.e. do we have an ask and use cases that warrant the separation?

Reply:
it also makes consumer code painful to write... if `canDefer()` might return `false` then the code-cleanliness advantages of having a `defer()` method are lost AFAICT.

Reply:
In general, I felt that the *drawback* of including `defer()` in the API was mostly that we had all of these extra comments about reference counting and so forth. It's easy enough to factor out the reference counting itself and make `Tracer` implementation easier, but we're still stuck with the complexity of the mental model.

Given that, I think we either (a) get rid of `canDefer()` and expect everything to be able to `defer()`, or (b) get rid of `defer()` and `Continuation` entirely.

Reply:
I think we can get rid of this if we chose the approach like https://github.com/opentracing/opentracing-java/pull/166/files#diff-6948b0b0cc49e4e8a33bb8c0b92ecfdcR30 when _ref conting_ can be plugged in at span creation time.

Reply:
I definitely like that ref counting approach - the only thing is that it doesn't make 'clear' the part about capturing state, as it all happens implicitly (through that `Observer` interface, that is).

However, as it seems there are more and more issues around the `defer` part, this may as well be an alternative (along with the wrapper approach suggested by @yurishkuro )

Reply:
I agree that `canDefer()` should go away.  I don't think we can remove `Continuation` though unless we have a similar concept to replace it with.